src/goto-harness/recursive_initialization.cpp

```diff
@@ -136,5 +151,180 @@ void recursive_initializationt::initialize_nondet(
   const recursion_sett &known_tags,
   code_blockt &body)
 {
-  body.add(code_assignt{lhs, side_effect_expr_nondett{lhs.type()}});
+  if(
+    lhs.id() != ID_symbol ||
+    !is_array_size_parameter(to_symbol_expr(lhs).get_identifier()))
+  {
+    body.add(code_assignt{lhs, side_effect_expr_nondett{lhs.type()}});
+  }
+}
+
+void recursive_initializationt::initialize_array(
+  const exprt &array,
+  std::size_t depth,
+  const recursion_sett &known_tags,
+  code_blockt &body)
+{
+  PRECONDITION(array.type().id() == ID_array);
+  const auto &array_type = to_array_type(array.type());
+  const auto array_size =
+    numeric_cast_v<std::size_t>(to_constant_expr(array_type.size()));
+  for(std::size_t index = 0; index < array_size; index++)
+  {
+    initialize(
+      index_exprt(array, from_integer(index, size_type())),
+      depth,
+      known_tags,
+      body);
+  }
+}
+
+bool recursive_initializationt::should_be_treated_as_array(
+  const irep_idt &array_name) const
+{
+  return initialization_config.pointers_to_treat_as_arrays.find(array_name) !=
+         initialization_config.pointers_to_treat_as_arrays.end();
+}
+
+bool recursive_initializationt::is_array_size_parameter(
+  const irep_idt &cmdline_arg) const
+{
+  return initialization_config.variables_that_hold_array_sizes.find(
+           cmdline_arg) !=
+         initialization_config.variables_that_hold_array_sizes.end();
+}
+
+optionalt<irep_idt> recursive_initializationt::get_associated_size_variable(
+  const irep_idt &array_name) const
+{
+  return optional_lookup(
+    initialization_config.array_name_to_associated_array_size_variable,
+    array_name);
+}
+
+void recursive_initializationt::initialize_dynamic_array(
```

Comment:
It looks like `initialization_config` and `goto_model` are the only member variables used in this function.

🤔 I wonder if it could be made free.

Reply:
Personally not fans of "free" functions, I don't consider them a sign of good design. Not going to do that, sorry.

Reply:
@LAJW You can make any function free, I'm just not sure what the point would be in this instance.

Reply:
As it is, this code tells me the entire class is necessary (which is not true - only 2 members are).

If you were to make it free, it would be unit-testable. And possible to reuse elsewhere. This code looks kind of generic. I think it's been done before.

If it was not bound to this particular context, it could be reused. But because it's a private method that adds something to a private field, that's not possible. So I guess that'll mean more comments like the one with which this function starts.

Reply:
As it is right now, this is a method of the recursive initialiser class, which is also tasked with initialising other things as well (pointers, strings,etc). If this were a free function, then it would be isolated it from its logical context, and not make a lot of sense, just to satisfy blindly a random design guideline which is not very appropriate in this case.

Let's please not look at the trees and lose sight of the forest.

Reply:
> As it is, this code tells me the entire class is necessary (which is not true - only 2 members are).

Basically I've been thinking about an approach to clean this up somewhat, however this would require significant rewrites of the entire functionality we have here so far, which I don't think makes sense to do as part of this PR.